lib/rubocop/cop/rspec/subject_stub.rb

```diff
@@ -74,71 +76,54 @@ class SubjectStub < Cop
         PATTERN
 
         def on_block(node)
+          # process only describe/context/... example groups
           return unless example_group?(node)
 
-          find_subject_stub(node) do |stub|
+          # skip already processed example group
+          # it's processed if is nested in one of the processed example groups
+          return unless (processed_example_groups & node.ancestors).empty?
+
+          # add example group to already processed
+          processed_example_groups << node
+
+          # find all custom subjects e.g. subject(:foo) { ... }
+          @named_subjects = find_all_named_subjects(node)
+
+          # look for method expectation matcher
+          find_subject_expectations(node) do |stub|
             add_offense(stub)
           end
         end
 
         private
 
-        # Find subjects within tree and then find (send) nodes for that subject
-        #
-        # @param node [RuboCop::Node] example group
-        #
-        # @yield [RuboCop::Node] message expectations for subject
-        def find_subject_stub(node, &block)
-          find_subject(node) do |subject_name, context|
-            find_subject_expectation(context, subject_name, &block)
-          end
+        def processed_example_groups
+          @processed_example_groups ||= Set[]
         end
 
-        # Find a subject message expectation
-        #
-        # @param node [RuboCop::Node]
-        # @param subject_name [Symbol] name of subject
-        #
-        # @yield [RuboCop::Node] message expectation
-        def find_subject_expectation(node, subject_name, &block)
-          # Do not search node if it is an example group with its own subject.
-          return if example_group?(node) && redefines_subject?(node)
-
-          # Yield the current node if it is a message expectation.
-          yield(node) if message_expectation?(node, subject_name)
-
-          # Recurse through node's children looking for a message expectation.
-          node.each_child_node do |child|
-            find_subject_expectation(child, subject_name, &block)
+        def find_all_named_subjects(node)
+          node.each_descendant(:block).each_with_object({}) do |child, h|
+            name = subject(child)
+            h[child.parent.parent] = name if name
           end
         end
 
-        # Check if node's children contain a subject definition
-        #
-        # @param node [RuboCop::Node]
-        #
-        # @return [Boolean]
-        def redefines_subject?(node)
-          node.each_child_node.any? do |child|
-            subject(child) || redefines_subject?(child)
+        def find_subject_expectations(node, subject_name = nil, &block)
```

Comment:
I find this code hard to reason about.
Did you remove `redefines_subject?` for the sake of performance?
Can't recall if we have some benchmarks for cops here. I agree with you that cops, especially those using `def_node_search` should be benchmarked. I'm guilty of making changes to certain cops that improve their correctness by sacrificing performance, but I was doing this blindfolded without knowing how much the performance suffers exactly.

Reply:
> Did you remove redefines_subject? for the sake of performance?

I just refactored the implementation. The case when subject is redefined is handled in another way.

Reply:
It should be fixed now